File: Keep_Stock_Web/services/imagem_storage.py
import os
import uuid

from supabase import create_client

SUPABASE_URL = "https://gysiqyxpmlxmlolqzaqs.supabase.co"

# Chave Anon (abaixo)
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imd5c2lxeXhwbWx4bWxvbHF6YXFzIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3NDk2MDcwNiwiZXhwIjoyMDkwNTM2NzA2fQ.DRqq9majrxnIT8oUnhrY2qQFB-OG6V4yOfTSGg27lpQ"

supabase = create_client(SUPABASE_URL,SUPABASE_KEY)

BUCKET = "Produtos"
BUCKET_CLIENTES = "Clientes"
BUCKET_FORNECEDORES = "Fornecedores"

EXTENSOES_PERMITIDAS = {
    ".jpg",
    ".jpeg",
    ".png",
    ".webp"
}


def upload_imagem(arquivo, bucket=BUCKET):

    if not arquivo or not arquivo.filename:
        return None

    extensao = os.path.splitext(arquivo.filename)[1].lower()

    # Formato permitidos
    if extensao not in EXTENSOES_PERMITIDAS:
        raise ValueError("Formato de imagem não permitido. " "Use JPEG, PNG ou WEBP.")

    nome_arquivo = f"{uuid.uuid4()}{extensao}"
    caminho = nome_arquivo
    arquivo_bytes = arquivo.read()

    # Caminho para o storage
    supabase.storage \
        .from_(bucket) \
        .upload(
            caminho,
            arquivo_bytes,
            {
                "content-type": arquivo.content_type
            }
        )

    url_publica = supabase.storage \
        .from_(bucket) \
        .get_public_url(caminho)

    return url_publica


def deletar_imagem(url, bucket=BUCKET):

    if not url:
        return

    try:
        prefixo = f"/storage/v1/object/public/{bucket}/"

        if prefixo not in url:
            print("URL de imagem inválida:", url)
            return

        caminho = url.split(prefixo, 1)[1]

        supabase.storage \
            .from_(bucket) \
            .remove([caminho])

    except Exception as e:
        print(f"Erro ao deletar imagem: {e}")

File: Mobile_Keep_Stock/src/services/imagemStorage.js
import { File } from 'expo-file-system';

import { supabase } from './supabase';

// Upload e remoção de imagens no Storage (espelha services/imagem_storage.py do sistema web)

export const BUCKET = 'Produtos';
export const BUCKET_CLIENTES = 'Clientes';
export const BUCKET_FORNECEDORES = 'Fornecedores';

const TIPOS_PERMITIDOS = {
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.png': 'image/png',
  '.webp': 'image/webp',
};

// Nome único do arquivo, no lugar do uuid4 usado no sistema web
function gerarNomeArquivo(extensao) {
  const bytes = crypto.getRandomValues(new Uint8Array(16));

  const identificador = Array.from(bytes, (byte) =>
    byte.toString(16).padStart(2, '0')
  ).join('');

  return `${identificador}${extensao}`;
}

function extrairExtensao(uri) {
  const caminho = uri.split('?')[0];
  const posicao = caminho.lastIndexOf('.');

  return posicao === -1 ? '' : caminho.slice(posicao).toLowerCase();
}

// Recebe a imagem no formato devolvido pelo seletor de imagens: { uri }
export async function uploadImagem(arquivo, bucket = BUCKET) {
  if (!arquivo || !arquivo.uri) return null;

  const extensao = extrairExtensao(arquivo.uri);
  const contentType = TIPOS_PERMITIDOS[extensao];

  // Formatos permitidos
  if (!contentType) {
    throw new Error('Formato de imagem não permitido. Use JPEG, PNG ou WEBP.');
  }

  const caminho = gerarNomeArquivo(extensao);
  const arquivoBytes = await new File(arquivo.uri).bytes();

  const { error } = await supabase.storage
    .from(bucket)
    .upload(caminho, arquivoBytes, { contentType });

  if (error) throw error;

  const { data } = supabase.storage.from(bucket).getPublicUrl(caminho);

  return data.publicUrl;
}

export async function deletarImagem(url, bucket = BUCKET) {
  if (!url) return;

  try {
    const prefixo = `/storage/v1/object/public/${bucket}/`;

    if (!url.includes(prefixo)) {
      console.log('URL de imagem inválida:', url);
      return;
    }

    const caminho = url.split(prefixo)[1];

    await supabase.storage.from(bucket).remove([caminho]);
  } catch (erro) {
    console.log(`Erro ao deletar imagem: ${erro}`);
  }
}
